pkg/decoder/decoder.go

```diff
@@ -7,102 +7,205 @@ package decoder
 
 import (
 	"bytes"
+	"regexp"
+	"time"
 
 	"github.com/DataDog/datadog-log-agent/pkg/config"
 	"github.com/DataDog/datadog-log-agent/pkg/message"
 )
 
-// Payload represents a list of bytes and an optional reference to its origin
-type Payload struct {
-	content []byte
-	offset  int64
-}
+// countDownDuration represents the time after which the buffered line is sent when there is no more incoming data
+const countDownDuration = 1 * time.Second
 
-// NewPayload returns a new decoder payload
-func NewPayload(content []byte, offset int64) *Payload {
-	return &Payload{content, offset}
-}
+// maxMessageLen represents the maximum length for a message
+var maxMessageLen = config.MaxMessageLen
 
-// Decoder splits raw data based on `\n`, and sends those messages to a channel
+// Decoder splits raw data into messages using '\n' for single-line logs and lineRe for multi-line logs
+// and sends them to a channel
 type Decoder struct {
 	InputChan  chan *Payload
 	OutputChan chan message.Message
+
 	msgBuffer  *bytes.Buffer
+	lineBuffer *bytes.Buffer
+
+	timer  *time.Timer
+	lineRe *regexp.Regexp
 }
 
 // InitializeDecoder returns a properly initialized Decoder
-func InitializedDecoder() *Decoder {
+func InitializeDecoder(source *config.IntegrationConfigLogSource) *Decoder {
```

Comment:
There are lots of if/else disjunctions, should we have more of a composition or inheritance pattern to avoid those and make the code more readable?

Reply:
Agreed